### PR TITLE
Loudspeak

### LINKED ISSUE #7119: Loud headsets for Heads of Staff
### Description

I am unsure of their name on /tg/, but they were headsets that made your radio chatter 2 times as bigger, thus making your messages more noticeable. (And more annoying)

### PR DESCRIPTION
**Pull request by MaxIsJoe**
closes #7119 

### Changelog

CL: [New] - When shouting by using "!!" your text will become larger just like on /tg/
CL: [New] - Headsets for command can now have the ability to toggle loudspeak which will make their text much bigger than the rest of the crew on comm channels.

### CODE REVIEW

**Comment by MaxIsJoe** [comment]
Up for review

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs [hunk 1]
```diff
@@ -83,7 +83,7 @@ public static void InvokeChatEvent(ChatEvent chatEvent)
 	/// Send a Chat Msg from a player to the selected Chat Channels
 	/// Server only
 	/// </summary>
-	public static void AddChatMsgToChat(ConnectedPlayer sentByPlayer, string message, ChatChannel channels)
+	public static void AddChatMsgToChat(ConnectedPlayer sentByPlayer, string message, ChatChannel channels, Loudness loudness)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: could default to normal loudness?

File: UnityProject/Assets/Scripts/Items/Headset.cs [hunk 2]
```diff
@@ -4,15 +4,28 @@
 /// <summary>
 ///     Headset properties
 /// </summary>
-public class Headset : NetworkBehaviour
+public class Headset : NetworkBehaviour, IInteractable<HandActivate>
 {
 	[SyncVar] public EncryptionKeyType EncryptionKey;
+	[SyncVar] public bool LoudSpeakOn = false;
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Items/Headset.cs:10]: Worth changing these to hooks

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs [hunk 3]
```diff
@@ -125,6 +125,22 @@ public static void AddChatMsgToChat(ConnectedPlayer sentByPlayer, string message
...
+				if (headset.LoudSpeakOn && channels.HasFlag(ChatChannel.Common) ||
+				    channels.HasFlag(ChatChannel.Command) || channels.HasFlag(ChatChannel.Security)
+				    || channels.HasFlag(ChatChannel.Engineering) || channels.HasFlag(ChatChannel.Medical)
+				    || channels.HasFlag(ChatChannel.Science)
+				    || channels.HasFlag(ChatChannel.Syndicate) || channels.HasFlag(ChatChannel.Supply))
+				{
+					chatEvent.VoiceLevel = headset.LoudspeakLevel;
+				}
+			}
+		}
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: DynamicItemStorage needs to be null checked as ghost, Ai and blob dont have storage

File: UnityProject/Assets/Scripts/Items/Headset.cs (hunk 2 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Items/Headset.cs:10]: Only need to do that if you want to do something on all clients when the value is changed, not needed here.

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs (hunk 3 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: and Item can be null

File: UnityProject/Assets/Scripts/Items/Headset.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Items/Headset.cs:10]: doesn't it constantly poll tho? if it's not a hook

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Items/Headset.cs:10]: dont think so

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Items/Headset.cs:10]: ok

**Comment by Aranclanos** [comment]
talking without a headset throws a runtime where bod pointed out and players are unable to speak, even OOC chat
as a client, I can edit the loudness level to anything I want and the server will accept it

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs (hunk 3 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: Done.

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: Alright now it's done done

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs [hunk 4]
```diff
@@ -121,9 +121,49 @@ public static void AddChatMsgToChat(ConnectedPlayer sentByPlayer, string message
...
+					if (headset != null)
+					{
+						if (headset.LoudSpeakOn && IsOnCorrectChannels())
+						{
+							chatEvent.VoiceLevel = headset.LoudspeakLevel;
+						}
+					}
+				}
+			}
+		}
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: Could you move this into its own function not a local one, its not needed inside and bloats it

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs [hunk 5]
```diff
@@ -121,9 +121,49 @@ public static void AddChatMsgToChat(ConnectedPlayer sentByPlayer, string message
...
+			{
+				return true;
+			}
+			return false;
+		}
+
+		//Check if is not a ghost/spectator and the player has an inventory.
+		void CheckVoiceLevel(PlayerScript script)
+		{
+			if (!script.mind.body.IsDeadOrGhost && script.DynamicItemStorage != null)
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Core/Chat/Chat.cs]:
```suggestion
			if (script.mind.body.IsDeadOrGhost == false && script.DynamicItemStorage != null)
```

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs (hunk 4 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: Means you can unnest it all too

**Review by Bod9001 — DISMISSED** [review]

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: Makes no sense for me to turn this into a whole proper function when it's never going to be used else where

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: Its just the function is now 134 lines long and needs reorganising

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: done

File: UnityProject/Assets/Scripts/Core/Chat/ChatEvent.cs [hunk 6]
```diff
@@ -79,6 +79,15 @@ public enum ChatModifier
 	Scotsman = 1 << 21
 }
 
+public enum Loudness
+{
+	QUITE,
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Core/Chat/ChatEvent.cs]:
```suggestion
	QUIET,
```

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Core/Chat/ChatEvent.cs]: done

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs [hunk 7]
```diff
@@ -121,9 +121,20 @@ public static void AddChatMsgToChat(ConnectedPlayer sentByPlayer, string message
 			speaker = (player == null) ? sentByPlayer.Username : player.playerName,
 			position = (player == null) ? TransformState.HiddenPos : player.PlayerChatLocation.AssumedWorldPosServer(),
 			channels = channels,
-			originator = (player == null) ? sentByPlayer.GameObject : player.PlayerChatLocation
+			originator = (player == null) ? sentByPlayer.GameObject : player.PlayerChatLocation,
+			VoiceLevel = loudness
 		};
 
+		//Check if is not a ghost/spectator and the player has an inventory.
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: Forgot to move this

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: ive done it, dw

**Review by ThatDan123 — APPROVED** [review]

**Review by corp-0 — APPROVED** [review]

**Pull request merged**